source/funkin/play/character/CharacterData.hx

```diff
@@ -281,41 +282,93 @@ class CharacterDataParser
    */
-  public static function getCharPixelIconAsset(char:String):String
+  public static function getCharPixelIcon(char:String):String
   {
-    var icon:String = char;
-
-    switch (icon)
+    var charPath = switch (char)
     {
-      case "bf-christmas" | "bf-car" | "bf-pixel" | "bf-holding-gf":
-        icon = "bf";
+      case "bf-christmas" | "bf-car" | "bf-pixel" | "bf-holding-gf" | "bf-dark":
+        "bfpixel";
       case "monster-christmas":
-        icon = "monster";
+        "monsterpixel";
       case "mom" | "mom-car":
-        icon = "mommy";
+        "mommypixel";
       case "pico-blazin" | "pico-playable" | "pico-speaker":
-        icon = "pico";
-      case "gf-christmas" | "gf-car" | "gf-pixel" | "gf-tankmen":
-        icon = "gf";
+        "picopixel";
+      case "gf-christmas" | "gf-car" | "gf-pixel" | "gf-tankmen" | "gf-dark":
+        "gfpixel";
       case "dad":
-        icon = "daddy";
+        "dadpixel";
       case "darnell-blazin":
-        icon = "darnell";
+        "darnellpixel";
       case "senpai-angry":
-        icon = "senpai";
+        "senpaipixel";
       case "spooky-dark":
-        icon = "spooky";
+        "spookypixel";
       case "tankman-atlas":
-        icon = "tankman";
+        "tankmanpixel";
+      case "pico-christmas" | "pico-dark":
+        "picopixel";
+      default:
+        '${char}pixel';
     }
 
-    var path = Paths.image("freeplay/icons/" + icon + "pixel");
-    if (Assets.exists(path)) return path;
+    if (!Assets.exists(Paths.image("freeplay/icons/" + charPath)))
+    {
+      trace('[WARN] Character ${char} has no freeplay icon.');
+      return null;
+    }
 
-    // TODO: Hardcode some additional behavior or a fallback.
-    return null;
+    return charPath;
+  }
+
+  /**
+   * Returns the idle frame of a character.
+   * @author TechnikTil
+   */
+  public static function getCharPixelIconAsset(char:String):FlxFrame
+  {
+    var charPath:String = "freeplay/icons/";
+
+    var convertChar:Null<String> = getCharPixelIcon(char);
+    if(convertChar != null)
+      return null;
+
+    charPath += convertChar;
+
+    var isAnimated = Assets.exists(Paths.file('images/$charPath.xml'));
+    var frame:FlxFrame = null;
+
+    if (isAnimated)
+    {
+      var frames = Paths.getSparrowAtlas(charPath);
+
+      var idleFrame:FlxFrame = frames.frames.find(function(frame:FlxFrame):Bool {
+        return frame.name.startsWith('idle');
+      });
+
+      if (idleFrame == null)
+      {
+        trace('[WARN] Character ${char} has no idle in their freeplay icon.');
+        return null;
+      }
+
+      // so, haxe.ui.backend.AssetsImpl uses the parent width and height, which makes the image go crazy when rendered
+      // so this is a work around so that it uses the actual width and height
+      var imageGraphic = flixel.graphics.FlxGraphic.fromFrame(idleFrame);
+
+      var imageFrame = flixel.graphics.frames.FlxImageFrame.fromImage(imageGraphic);
```

Comment:
This causes extra lag for the engine, so I made a PR to fix the issue altogether.
haxeui/haxeui-flixel#53

Reply:
(this has been merged in haxeui-flixel now fyi 👍 )

Reply:
yippee :D